### ISSUE TITLE
Adjust Search (v0/search) API call in vets-api to properly handle Search.gov maintenance windows

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-cms#16784]
https://dsva.slack.com/archives/C52CL1PKQ/p1704842021309269

On 1/9/24, Search.gov had a planned maintenance window we were not aware of. 

## Impact
When Search.gov went down, we saw 500s on the `v0/search` endpoint, leading to the below error message. When Search.gov has planned maintenance windows, their SOP is to return a "no results" message (see below screenshot). vets-api handles this as an internal service error and produces the below failure. Instead, we need to plan for 5 distinct scenarios:
- Successful call with results returned from Search.gov
- Successful call with no results for search term from Search.gov
- Successful call with "no results" (maintenance window) from Search.gov
- Failed call to Search.gov (successful vets-api response)
- Failed call to vets-api endpoint

Search.gov status from 1/9/24
<img width="500" alt="Screenshot 2024-01-09 at 5 23 08 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/19175324/3dbfabca-81ae-449f-8d6d-033c68082197">

### Site visitors get an error
Attempting to search on VA.gov returned an error response: "Your search didn't go through:
We’re sorry. Something went wrong on our end, and your search didn't go through. Please try again."

![Screenshot 2024-01-09 at 3 31 11 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/b890b8b9-9983-4a20-b888-58935e3975dc)

### Monitor alarms
The following monitors were affected: 
   1. [VA.gov search success rate below threshold](https://vagov.ddog-gov.com/monitors/169139?from_ts=1704841679000&to_ts=1704842879000&source=monitor_notif)
   3. [VA.gov homepage load is failing](https://vagov.ddog-gov.com/monitors/169094?from_ts=1704841514000&to_ts=1704842714000&source=monitor_notif) -- uses a synthetic browser test, [VA.gov homepage functional browser test](https://vagov.ddog-gov.com/synthetics/details/gsg-zui-hg8), which also runs a search synthetic test, which failed and then failed our monitor

## Goal
Per #13736 , Search is already configured and able to use downtime notifications. But: we need to figure out: how to be aware of Search.gov planned maintenance in advance, in order to take advantage of that and set up the downtime notice.

## ACs
- [x] Successful call with results returned from Search.gov: Search returns as expected (current behavior)
- [x] Successful call with no results for search term from Search.gov: Failed call indicates server error 
- [ ] Successful call with "no results" (maintenance window) from Search.gov: Should send back a code other than success and failure; Message should be ~ Search.gov is down 
- [x] Failed call to Search.gov (successful vets-api response): Failed call indicates server error 
- [x] Failed call to vets-api endpoint; Failed call indicates server error

**Comment by jilladams** [comment]
Related slack summary: https://dsva.slack.com/archives/C52CL1PKQ/p1704906464546279

**Comment by jilladams** [comment]
Search.gov response re: their downtime and why we got 500s in this case: 

> The windows on Tues/Thurs are our regular production push windows. It's rare that we don't do a push of some sort on these days. These pushes include regular maintenance (e.g., updates or upgrades, bug fixes) and incremental feature sub-component releases. When an app is deployed to a given host, there is a chance that that host will return an error if the load balancer happens to send a query to that host in that moment. We've looked into whether we can exclude the host being deployed to from the load balancer, and it's difficult.

> If you see a larger maintenance window, or one at an unusual time, like this one, it's because we're doing a deep change that needs to be handled carefully and separately from a regular deploy. This change was to the TLS settings for our internal network, and will not be happening again, now that it's done.

> Regarding the 500s you received - could you provide the timeframe (start and end)? I'm not sure why you got 500's and not 'no results,' but it's likely due to the way the results API queries the system, as opposed to our hosted results page. Apologies for that.

> For unplanned downtime, our SOP is to post to our status page right away, and then send an email to customers if the outage extends beyond the 30 minute mark. In this case, knowing we were expecting limited downtime, we planned for outside core business hours and just posted to the status page, but your request for advance notice is heard!

**Comment by jilladams** [comment]
New wrinkle probably for follow up with Search.gov support: 
https://dsva.slack.com/archives/C05THHJHH2R/p1704914590248559?thread_ts=1704908199.722959&cid=C05THHJHH2R
> it looks like we were actually getting 504 gateway timeouts from them, example: [https://vagov.ddog-gov.com/apm/resource/search/rack.request/72b7d62abdbc2323?query=env[…]iew=spans&start=1704841350056&end=1704841448009&paused=true](https://vagov.ddog-gov.com/apm/resource/search/rack.request/72b7d62abdbc2323?query=env%3Aeks-prod%20operation_name%3Arack.request%20resource_name%3A%22V0%3A%3ASearchController%23index%22%20service%3Asearch&env=eks-prod&graphType=flamegraph&shouldShowLegend=true&sort=time&spanID=656414702749308482&timeHint=1704841407297&topGraphs=latency%3Alatency%2Chits%3Acount%2Cerrors%3Acount%2CbreakdownAs%3Apercentage&trace=AgAAAYzwd2tBPj2vigAAAAAAAAAYAAAAAEFZendkM0VEQUFDXy1uNWM4RjYzb3ZBbQAAACQAAAAAMDE4Y2YwYWItMmVjZS00YzRiLTk4ZjgtNWI3NzhhZGM3ZmNl&traceID=2805929711136253874&traces=qson%3A%28data%3A%28%29%2Cversion%3A%210%29&view=spans&start=1704841350056&end=1704841448009&paused=true)

![Screenshot 2024-01-10 at 11 42 43 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/85581471/16452531-32cd-4e3e-92c3-ec1885b82046)

@randimays do you mind sending that info to Search.gov support and cc me / Fran?

**Comment by randimays** [comment]
Sent a new reply to Search.gov and added @jilladams and @FranECross on that email thread.

**Comment by jilladams** [comment]
Platform Backend COP owns vets-api.

**Comment by Becapa** [comment]
I think in the ACs above, the only one that doesn't appear to already have something in place is AC#3 (api calls during the maintenance window). Checking for a 504 seems pretty straightforward, but according to the conversation mentioned above in Slack, 504 errors shouldn't happen again. Searching through the [GSA search-gov codebase](https://github.com/GSA/search-gov), I was unable to find exactly how the response is structured when we make a call to the API during the maintenance window. There also isn't any mention of it on the [Search.gov Results API site](https://open.gsa.gov/api/searchgov-results/). I've sent an email to Search.gov support asking what, exactly, we should be getting as a response when we hit the API during a maintenance window.

**Comment by jilladams** [comment]
Talked to Michael - we concurred that Search.gov may not give us any better information, based on prior experience with their support. So: we'll add handling for the 504 error code. (Jordan has already done copy, and once this is done, the FE ticket will be unblocked to proceed.)

**Comment by jilladams** [comment]
Noting: Today we saw a traffic dip on search, with a bump up in 503 errors. 

[Example response](https://vagov.ddog-gov.com/apm/services/search/operations/rack.request/errors?dependencyMap=qson%3A%28data%3A%28dependencyLayout%3AhybridView%29%2Cversion%3A%210%29&env=eks-prod&errors=qson%3A%28data%3A%28issueSort%3ATOTAL_COUNT%2CtopN%3A%215%29%2Cversion%3A%211%29&issueId=f1d4ac4e-9eb4-11ee-88e7-da7ad0900007&panels=qson%3A%28data%3A%28%29%2Cversion%3A%210%29&summary=qson%3A%28data%3A%28visible%3A%21t%2Cerrors%3A%28selected%3Acount%29%2Chits%3A%28selected%3Acount%29%2Clatency%3A%28selected%3Alatency%2Cslot%3A%28agg%3A95%29%2Cdistribution%3A%28isLogScale%3A%21f%29%2CshowTraceOutliers%3A%21f%29%2Csublayer%3A%28slot%3A%28layers%3Aservice%29%2Cselected%3Apercentage%29%29%2Cversion%3A%211%29&view=spans&start=1706290140000&end=1706290260000&paused=true):   
> NoMethodError: undefined method dig' for "<h2>This website is under heavy load (queue full)</h2><p>We're sorry, too many people are accessing this website at the same time. We're working on this problem. Please try again later.</p>"

@Becapa will 503 errors also get handled in the work you've done in this ticket already? If not, what would the lift be to add that?

**Comment by Becapa** [comment]
Response from Dawn McCleskey (Search.gov Support):

> Hello Michael, Our hosted results page displays a no results message if there are zero results. Via the API, it would be a 200 and look like this: https://api.gsa.gov/technology/searchgov/v2/results/i14y?affiliate=usasearch_sandbox&access_key=s2nEFodS5_IsIsm8n0isyFsDPrNoBYgmtehqpzmzgP8=&query=aslivnaoinvaoivn
>  
> During the maintenance on January 9, we expected the app to report zero results, but unfortunately in VA's case you were getting 5xx responses. We have not had reports from other API users about this time period, so I don't know what sorts of responses they were getting at the time.
>  
> We do not have any more deep networking work like we needed to do on Jan 9, but we have a plan going forward if we do need to make this sort of deep change where we anticipate degradation, to make changes to our traffic distribution to minimize interruptions to the extent possible.

So basically, it doesn't differentiate between being in maintenance or a specific search not having results. I think this backs up that we should just be looking at the 500 errors and handling those.

**Comment by jilladams** [comment]
@Becapa feel free to ignore this thread - consider yourself done on this ticket. 

From scrum today: 
* In order to close this ticket: the PR notes that the tests recreate failure scenarios in order to verify the changes, so: tests pass and we can say this is done. 
* There are more and different questions that follow this change, that are mystical and we need to understand in order to know if: 
  * We need more API work for more and different kinds of failures, or if this change is sufficient.
  * We have sufficient detail coming through this change to the Front-end, to display better error messages via https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16801

Putting these notes here for now, for the sake of having them somewhere logical for discussion in FE refinement tomorrow. We may opt to close this ticket anyway, and move follow on conversation somewhere else, TBD. 

@randimays @dsasser @FranECross my notes / questions below for tomorrow's FE refinement discussion: 

## Jill's string wall brain about random Search monitoring failures
There is a lot of search noise in monitoring. I don't know how to interpret it, if this change helps or affects it or not, etc. Here is the dump of things / questions I have, for discussion: 

### Current ticket / PR / next steps
* In this PR: https://github.com/department-of-veterans-affairs/vets-api/pull/15337/files, we add 503 and 504 handling. 
  * Will the error that pushes through be clear to FE whether it's coming from upstream/Search.gov vs. vets-api? Is that possible to make clear?
  * The messages provided in there: is that text required to come back in the API response, in order to flag it as a 503 / 504 upstream response? 

```      
      SEARCH_503:
        <<: *external_defaults
        title: Service Unavailable
        code: 'SEARCH_503'
        detail: 'Search.gov service is currently unavailable'
        status: 503
      SEARCH_504:
        <<: *external_defaults
        title: Gateway Timeout
        code: 'SEARCH_504'
        detail: 'Did not receive a timely response from Search.gov'
        status: 504
```
Unrelated, but a question from that PR: what the heck do vcr cassettes have to do with anything??

### On the monitoring side
**1. Last night's 503/504s**: https://dsva.slack.com/archives/C05THHJHH2R/p1707243766443409
We got 17 503s with a log message: `Breakers::OutageException: Outage detected on Search/Results beginning at 1707208317`
* Is that coming from vets-api? From search.gov? How can we know? (this is maybe an existential question?)

**2. The Phusion / unknown error:** https://dsva.slack.com/archives/C05THHJHH2R/p1706565368152289?thread_ts=1706564391.025529&cid=C05THHJHH2R
`NoMethodError: undefined method dig' for "<h2>This website is under heavy load (queue full)</h2><p>We're sorry, too many people are accessing this website at the same time. We're working on this problem. Please try again later.</p>":String`
* Will the change made in this PR flag that correctly as an upstream error? Or bc the message is different than the code block above, no, and it needs its own handling? (Do we really have to have an error handling case for every possible API response error message we could ever get?)
* Is it actually an upstream error? How could we figure that out? Does this need its own ticket to investigate? Adrian thinks it comes from Search.gov and we should handle it as an upstream error. 
  * ("I don't see any place we're using phusion either so fairly confident that it's coming from their backend. In the end I think it's just another instance of "the [search.gov](http://search.gov/) API is not as reliable as we'd like it to be". It's probably worth a ticket to update the controller to anticipate this error case and report it as an upstream error rather than throwing what looks like a vets-api error.")

**Comment by FranECross** [comment]
Per Jill's note above, I'm closing this ticket, and additional ticket(s) will be written to take into consideration new information regarding search.gov, etc.
